### PR TITLE
GC core impl improvements

### PR DESCRIPTION
**Pull request by sinui0**
Some things included in the scope of this PR:

- [x] Incorporate interface changes from #51 (will change base to dev after this is merged)
- [x] Add newtypes to enforce safety and consistency invariants
- [x] Reorganize label structs into it's own module
- [x] Incorporate changes into DualEx impl
- [x] Refactor msg types and corresponding conversions

### CODE REVIEW

File: mpc-core/src/garble/evaluator.rs [hunk 1]
```diff
@@ -2,57 +2,53 @@ use cipher::{consts::U16, BlockCipher, BlockEncrypt};
...
 use mpc_circuits::{Circuit, Gate};
 
-use super::{
-    circuit::{BinaryLabel, SanitizedInputLabels},
-    EncryptedGate, GarbledCircuit,
-};
-
 /// Evaluates AND gate
 #[inline]
 pub(crate) fn and_gate<C: BlockCipher<BlockSize = U16> + BlockEncrypt>(
```
Comment (th4s) [inline mpc-core/src/garble/evaluator.rs:11]: Is this the half-gate method? Can we add a comment?

File: mpc-core/src/garble/label.rs [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+    }
+
+    /// Generates pairs of wire labels \[W_0, W_0 ^ delta\]
+    pub fn generate<R: Rng + CryptoRng>(
+        rng: &mut R,
+        delta: Option<Delta>,
+        count: usize,
+        offset: usize,
+    ) -> (Vec<Self>, Delta) {
+        let delta = match delta {
```
Comment (th4s) [inline mpc-core/src/garble/label.rs:125]: Use `unwrap_or_else`?

File: mpc-core/src/msgs/garble.rs [hunk 3]
```diff
@@ -1,10 +1,330 @@
...
+                        .map(OutputEncoding::from)
+                        .collect::<Vec<OutputEncoding>>(),
+                )
+            }),
+        }
+    }
+}
+
+impl crate::garble::GarbledCircuit<garble::Partial> {
+    pub fn from_msg(circ: Arc<Circuit>, msg: GarbledCircuit) -> Result<Self, crate::garble::Error> {
```
Comment (th4s) [inline mpc-core/src/msgs/garble.rs:88]: Is this too computation intensive for a `TryFrom` implementation? Does it make sense to break it down into smaller pieces?

**Review by th4s — COMMENTED** [review]
:+1: lgtm

File: mpc-core/src/msgs/garble.rs (hunk 3 above)
Comment (th4s) [reply mpc-core/src/msgs/garble.rs:88]: Hmm now that I see it, I have changed my mind, lol. I agree, it is too clunky, and it does not make sense to create a wrapper type for both structs just to satisfy the `TryFrom` interface. Probably the best to leave it like it is.

File: mpc-core/src/garble/label.rs [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+    }
+}
+
+/// Input labels that have been sanitized are safe to use to evaluate a garbled circuit
+///
+/// It is important to check that the generator has provided the expected input labels,
+/// otherwise they may have an opportunity to behave maliciously to extract the evaluator's
+/// private inputs.
+#[derive(Debug, Clone)]
+pub struct SanitizedInputLabels(Vec<WireLabel>);
```
Comment (themighty1) [inline mpc-core/src/garble/label.rs:271]: Could we please have tests which trigger all the error conditions, since this code is security-critical.

**Review by themighty1 — CHANGES_REQUESTED** [review]
nicely done, left some comments.

**Review by th4s — APPROVED** [review]
nice :rocket:

**Pull request merged**